File: pkg/utils/vec3.go
package utils

import (
	"math"
)

// Vec3 is a 3D vector.
type Vec3 struct {
	D [3]float64
}

// NewVec3 creates a new Vec3.
func NewVec3(x, y, z float64) Vec3 {
	return Vec3{[3]float64{x, y, z}}
}

// Add adds the given vector to this vector
// and returns the result.
func (v Vec3) Add(arg Vec3) Vec3 {
	return NewVec3(v.D[0]+arg.D[0], v.D[1]+arg.D[1], v.D[2]+arg.D[2])
}

// Sub subtracts the given vector from this vector
// and returns the result.
func (v Vec3) Sub(arg Vec3) Vec3 {
	return NewVec3(v.D[0]-arg.D[0], v.D[1]-arg.D[1], v.D[2]-arg.D[2])
}

// Mul multiplies the vector with the given argument
// and returns the result.
func (v Vec3) Mul(arg float64) Vec3 {
	return NewVec3(v.D[0]*arg, v.D[1]*arg, v.D[2]*arg)
}

// Div divides the vector with the given argument
// and returns the result.
func (v Vec3) Div(arg float64) Vec3 {
	arg = 1 / arg
	return NewVec3(v.D[0]*arg, v.D[1]*arg, v.D[2]*arg)
}

// Dot calculates the dot product of this vector with the given vector.
func (v Vec3) Dot(arg Vec3) float64 {
	return v.D[0]*arg.D[0] + v.D[1]*arg.D[1] + v.D[2]*arg.D[2]
}

// DotSelf returns the dot product of the vector with itself.
// It is equivalent to its magnitude squared.
func (v Vec3) DotSelf() float64 {
	return v.Dot(v)
}

// Cross calculates the cross product of this vector with the given vector
// and returns the result.
func (v Vec3) Cross(arg Vec3) Vec3 {
	return NewVec3(
		v.D[1]*arg.D[2]-v.D[2]*arg.D[1],
		v.D[2]*arg.D[0]-v.D[0]*arg.D[2],
		v.D[0]*arg.D[1]-v.D[1]*arg.D[0],
	)
}

// Mag calculates the magnitude of the vector.
func (v Vec3) Mag() float64 {
	return math.Sqrt(v.DotSelf())
}

// Dir calculates the direction (or unit vector) of this vector.
func (v Vec3) Dir() Vec3 {
	return v.Div(v.Mag())
}

// ToColour converts this vector to a Colour type by mapping
// the x, y, z values to r, g, b values respectively.
func (v Vec3) ToColour() Colour {
	return Colour(v)
}

// Reflected calculates and returns the reflection of this vector
// for the given normal.
//
// To understand the formula, go to -
// https://raytracing.github.io/books/RayTracingInOneWeekend.html#metal/mirroredlightreflection
func (v Vec3) Reflected(normal Vec3) Vec3 {
	return v.Sub(normal.Mul(v.Dot(normal) * 2))
}

// Refracted calculates and returns the refraction of this vector
// for the given normal and refractive-index-ratio.
//
// The second parameter (named "rir") is the refractive-index-ratio.
// It is equal to the refractive index of the destination material divided
// by the refractive index of the source material.
//
// For more information, go to -
// https://raytracing.github.io/books/RayTracingInOneWeekend.html#dielectrics/snell'slaw
func (v Vec3) Refracted(normal Vec3, rir float64) Vec3 {
	// These are the perpendicular and parallel components of the refracted ray respectively.
	// The final result will be the addition of these components.
	var perpendicular, parallel Vec3

	// Calculating the incidence direction beforehand to save calculation time.
	vDir := v.Dir()
	// Cosine of the angle between the incident ray and the normal.
	cosine := vDir.Mul(-1).Dot(normal)
	// Value of cosine should never exceed 1.
	cosine = math.Min(cosine, 1)

	// Calculate the components.
	perpendicular = vDir.Add(normal.Mul(cosine)).Mul(rir)
	parallel = normal.Mul(-math.Sqrt(math.Abs(1 - perpendicular.DotSelf())))

	// Get the final result by adding the two components.
	return perpendicular.Add(parallel)
}

// Lerp stands for Linear Interpolation.
//
// The formula for linear interpolation is given by:
// final = (1 - x) * start + x * end.
func (v Vec3) Lerp(end Vec3, x float64) Vec3 {
	return v.Mul(1 - x).Add(end.Mul(x))
}

// IsNearZero returns true if ALL components of the vector are "very" close to zero.
func (v Vec3) IsNearZero() bool {
	precision := 0.00001
	return v.D[0] < precision && v.D[1] < precision && v.D[2] < precision
}

File: src/vec3.rs
/// Vec3 is a 3D vector.
#[derive(Clone, Copy)]
pub struct Vec3 {
    comp: [f32; 3],
}

impl Vec3 {
    /// Constructor.
    pub fn new(x: f32, y: f32, z: f32) -> Self {
        Vec3 { comp: [x, y, z] }
    }

    /// mag returns the magnitude/length of the vector.
    pub fn mag(&self) -> f32 {
        (self.x() * self.x() + self.y() * self.y() + self.z() * self.z()).sqrt()
    }

    /// dir returns the corresponding unit vector.
    pub fn dir(&self) -> Vec3 {
        *self / self.mag()
    }

    /// dot product with another vector.
    pub fn dot(&self, other: &Vec3) -> f32 {
        self.x() * other.x() + self.y() * other.y() + self.z() * other.z()
    }

    /// dot_self returns dot product with itself.
    pub fn dot_self(&self) -> f32 {
        self.dot(self)
    }

    /// cross product with another vector.
    pub fn cross(&self, other: &Vec3) -> Vec3 {
        Vec3::new(
            self.y() * other.z() - self.z() * other.y(),
            self.z() * other.x() - self.x() * other.z(),
            self.x() * other.y() - self.y() * other.x(),
        )
    }

    /// x component of the vector.
    #[inline]
    pub fn x(&self) -> f32 {
        self.comp[0]
    }

    /// y component of the vector.
    #[inline]
    pub fn y(&self) -> f32 {
        self.comp[1]
    }

    /// z component of the vector.
    #[inline]
    pub fn z(&self) -> f32 {
        self.comp[2]
    }
}

/// Operator overload for addition.
impl std::ops::Add<Vec3> for Vec3 {
    type Output = Vec3;

    fn add(self, arg: Vec3) -> Vec3 {
        Vec3::new(self.x() + arg.x(), self.y() + arg.y(), self.z() + arg.z())
    }
}

/// Operator overload for negation.
impl std::ops::Neg for Vec3 {
    type Output = Vec3;

    fn neg(self) -> Vec3 {
        Vec3::new(-self.x(), -self.y(), -self.z())
    }
}

/// Operator overload for subtraction.
impl std::ops::Sub<Vec3> for Vec3 {
    type Output = Vec3;

    fn sub(self, arg: Vec3) -> Vec3 {
        self + (-arg)
    }
}

/// Operator overload for mutliplication.
impl std::ops::Mul<f32> for Vec3 {
    type Output = Vec3;

    fn mul(self, arg: f32) -> Vec3 {
        Vec3::new(self.x() * arg, self.y() * arg, self.z() * arg)
    }
}

/// Operator overload for division.
impl std::ops::Div<f32> for Vec3 {
    type Output = Vec3;

    fn div(self, arg: f32) -> Vec3 {
        self * (1.0 / arg)
    }
}
